### PR TITLE
Improve: K&R style for if-then-else

### LINKED ISSUE #777: java style ite
it would be very nice if `parens-ite` could have a mode where instead of:

```ocaml
if x1 then foo1
else if x2 then (
  foo2;
  bar2 )
else (
  foo3 )
```
it would print

```ocaml
if x1 then foo1
else if x2 then (
  foo2;
  bar2
) else (
  foo3
)
```

that is, the closing paren is on its own line, but can be followed by `else` (like braces in the standard java style).

### PR DESCRIPTION
**Pull request by gpetiot**
Fix #777

### CODE REVIEW

File: test/passing/ite_kr.ml.ref [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+    ) else (
+      some other action
+    )
+  ) else (
+    some default action
+  )
+
+let foo =
+  if cmp < 0 then (
+    (* foo *)
```
Comment (gpetiot) [inline test/passing/ite_kr.ml.ref]: Originlaly this comment was floating on the right of the `then` keyword, don't know what to do with it.

File: src/Fmt_ast.ml [hunk 2]
```diff
@@ -1749,6 +1749,25 @@ and fmt_expression c ?(box = true) ?pro ?epi ?eol ?parens ?(indent_wrap = 0)
                           )
                       $ fmt_if parens_bch (if imd then " )" else ")") )
                     $ fmt_if (not last) "@ "
+                | `K_R ->
```
Comment (Julow) [inline src/Fmt_ast.ml]: A lot of code could be factorized (eg. the `xcnd` part). What about using the `Params` pattern here ?

Comment (gpetiot) [reply src/Fmt_ast.ml]: Done, but not sure it's easier to read

**Comment by gpetiot** [comment]
> How hard would it be to remove the parenthesis if both expressions are "simple" ?

Done

File: src/Fmt_ast.ml (hunk 2 above)
Comment (Julow) [reply src/Fmt_ast.ml]: Not sure too. But I think it's a lot more maintainable now.

File: src/Fmt_ast.ml [hunk 3]
```diff
@@ -1744,91 +1744,22 @@ and fmt_expression c ?(box = true) ?pro ?epi ?eol ?parens ?(indent_wrap = 0)
...
-                        )
-                    $ fmt_if (not last) "@ " )))
+                Params.get_if_then_else c.conf ~first ~last ~parens
+                  ~parens_bch ~xcond:xcnd ~expr_loc:pexp_loc
+                  ~fmt_extension_suffix:(fmt_extension_suffix c ext)
+                  ~fmt_attributes:
+                    (fmt_attributes c ~pre:(fmt " ") ~key:"@"
+                       pexp_attributes)
+                  ~fmt_cond:(fmt_expression c)
+                |> fun (p : Params.if_then_else) ->
```
Comment (Julow) [inline src/Fmt_ast.ml]: Why not `let p = .. in` ?

File: test/passing/ite_kr.ml.ref (hunk 1 above)
Comment (Julow) [reply test/passing/ite_kr.ml.ref]: I think breaking when there is parenthesis is fine. In this case I have no opinion.

File: src/Fmt_ast.ml (hunk 3 above)
Comment (gpetiot) [reply src/Fmt_ast.ml]: I don't remember either why I wrote it like this for the other calls to `Params.*`.

**Comment by gpetiot** [comment]
test_branch diff for the compact mode:  (the old lines violated the 100 chars margin)
```ocaml
diff --git a/infer/src/biabduction/Dom.ml b/infer/src/biabduction/Dom.ml
index e5ada3a3a..d64968995 100644
--- a/infer/src/biabduction/Dom.ml
+++ b/infer/src/biabduction/Dom.ml
@@ -1014,7 +1014,8 @@ let rec exp_partial_join (e1 : Exp.t) (e2 : Exp.t) : Exp.t =
       in
       Exp.BinOp (Binop.PlusPI, e1'', e2'')
   | Exp.BinOp (binop1, e1, e1'), Exp.BinOp (binop2, e2, e2') ->
-      if not (Binop.equal binop1 binop2) then ( L.d_strln "failure reason 24" ; raise Sil.JoinFail )
+      if not (Binop.equal binop1 binop2) then (
+        L.d_strln "failure reason 24" ; raise Sil.JoinFail )
       else
         let e1'' = exp_partial_join e1 e2 in
         let e2'' = exp_partial_join e1' e2' in
@@ -1023,7 +1024,8 @@ let rec exp_partial_join (e1 : Exp.t) (e2 : Exp.t) : Exp.t =
       if not (Pvar.equal pvar1 pvar2) then ( L.d_strln "failure reason 25" ; raise Sil.JoinFail )
       else e1
   | Exp.Lfield (e1, f1, t1), Exp.Lfield (e2, f2, _) ->
-      if not (Typ.Fieldname.equal f1 f2) then ( L.d_strln "failure reason 26" ; raise Sil.JoinFail )
+      if not (Typ.Fieldname.equal f1 f2) then (
+        L.d_strln "failure reason 26" ; raise Sil.JoinFail )
       else Exp.Lfield (exp_partial_join e1 e2, f1, t1) (* should be t1 = t2 *)
   | Exp.Lindex (e1, e1'), Exp.Lindex (e2, e2') ->
       let e1'' = exp_partial_join e1 e2 in
@@ -1115,7 +1117,8 @@ let rec exp_partial_meet (e1 : Exp.t) (e2 : Exp.t) : Exp.t =
       if not (Unop.equal unop1 unop2) then ( L.d_strln "failure reason 31" ; raise Sil.JoinFail )
       else Exp.UnOp (unop1, exp_partial_meet e1 e2, topt1) (* should be topt1 = topt2 *)
   | Exp.BinOp (binop1, e1, e1'), Exp.BinOp (binop2, e2, e2') ->
-      if not (Binop.equal binop1 binop2) then ( L.d_strln "failure reason 32" ; raise Sil.JoinFail )
+      if not (Binop.equal binop1 binop2) then (
+        L.d_strln "failure reason 32" ; raise Sil.JoinFail )
       else
         let e1'' = exp_partial_meet e1 e2 in
         let e2'' = exp_partial_meet e1' e2' in
@@ -1130,7 +1133,8 @@ let rec exp_partial_meet (e1 : Exp.t) (e2 : Exp.t) : Exp.t =
       if not (Pvar.equal pvar1 pvar2) then ( L.d_strln "failure reason 35" ; raise Sil.JoinFail )
       else e1
   | Exp.Lfield (e1, f1, t1), Exp.Lfield (e2, f2, _) ->
-      if not (Typ.Fieldname.equal f1 f2) then ( L.d_strln "failure reason 36" ; raise Sil.JoinFail )
+      if not (Typ.Fieldname.equal f1 f2) then (
+        L.d_strln "failure reason 36" ; raise Sil.JoinFail )
       else Exp.Lfield (exp_partial_meet e1 e2, f1, t1) (* should be t1 = t2 *)
   | Exp.Lindex (e1, e1'), Exp.Lindex (e2, e2') ->
       let e1'' = exp_partial_meet e1 e2 in
diff --git a/infer/src/integration/Driver.ml b/infer/src/integration/Driver.ml
index 527f70260..1c25018f1 100644
--- a/infer/src/integration/Driver.ml
+++ b/infer/src/integration/Driver.ml
@@ -278,7 +278,8 @@ let capture ~changed_files = function
               updated_buck_cmd )
             else build_cmd ) )
       in
-      if in_buck_mode && Config.flavors then ( RunState.set_merge_capture true ; RunState.store () ) ;
+      if in_buck_mode && Config.flavors then (
+        RunState.set_merge_capture true ; RunState.store () ) ;
       run_command ~prog:infer_py ~args
         ~cleanup:(function
           | Error (`Exit_non_zero exit_code)
```

**Pull request merged**